types/wlr_pointer_constraints_v1.c

```diff
@@ -0,0 +1,387 @@
+#include <assert.h>
+#include <stdbool.h>
+#include <stdlib.h>
+#include <pixman.h>
+#include <wayland-server.h>
+#include <wlr/types/wlr_box.h>
+#include <wlr/types/wlr_pointer_constraints_v1.h>
+#include <wlr/types/wlr_region.h>
+#include <wlr/util/log.h>
+
+#include "util/signal.h"
+#include "pointer-constraints-unstable-v1-protocol.h"
+
+static const struct zwp_locked_pointer_v1_interface locked_pointer_impl;
+static const struct zwp_confined_pointer_v1_interface confined_pointer_impl;
+static const struct zwp_pointer_constraints_v1_interface pointer_constraints_impl;
+
+static struct wlr_pointer_constraint_v1 *pointer_constraint_from_resource(
+		struct wl_resource *resource) {
+	assert(
+		wl_resource_instance_of(
+			resource, &zwp_confined_pointer_v1_interface,
+				&confined_pointer_impl) ||
+		wl_resource_instance_of(
+			resource, &zwp_locked_pointer_v1_interface,
+				&locked_pointer_impl));
+	return wl_resource_get_user_data(resource);
+}
+
+static struct wlr_pointer_constraints_v1 *pointer_constraints_from_resource(
+		struct wl_resource *resource) {
+	assert(wl_resource_instance_of(resource, &zwp_pointer_constraints_v1_interface,
+		&pointer_constraints_impl));
+	return wl_resource_get_user_data(resource);
+}
+
+// Thankfully zwp_confined_pointer_v1_send_{un,}confined work on both locked and confined pointer constraints
+static void send_constrained(struct wlr_pointer_constraint_v1 *constraint) {
+	constraint->active = true;
+
+	wlr_log(L_DEBUG, "constrained %p", constraint);
+	zwp_confined_pointer_v1_send_confined(constraint->resource);
+}
+static void send_unconstrained(struct wlr_pointer_constraint_v1 *constraint) {
+	if (!constraint->active) {
+		return;
+	}
+	constraint->active = false;
+
+	wlr_signal_emit_safe(
+		&constraint->pointer_constraints->events.constraint_inactive,
+		constraint);
+
+	wlr_log(L_DEBUG, "unconstrained %p", constraint);
+	zwp_confined_pointer_v1_send_unconfined(constraint->resource);
+	if (constraint->lifetime == ZWP_POINTER_CONSTRAINTS_V1_LIFETIME_ONESHOT) {
+		wl_resource_destroy(constraint->resource);
```

Comment:
The server shouldn't destroy client resources on his own, this will make clients crash

Reply:
How should the object be disposed of then? It doesn't create any problems with my example clients, but perhaps it does if I attempt to use the pointer constraint again after it's destroyed, which shouldn't be possible, no?

Reply:
The client destroys the object when it's done with it, with the `destroy` request.

Reply:
Then if a client has to monitor for when it leaves the surface to destroy it any way, what's the purpose of oneshot?

Reply:
Read the protocol. It's written that the client should destroy the object when it receives `unconfined`.